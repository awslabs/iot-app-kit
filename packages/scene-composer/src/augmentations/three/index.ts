export * from './Anchor';
export * from './Viewpoint';
