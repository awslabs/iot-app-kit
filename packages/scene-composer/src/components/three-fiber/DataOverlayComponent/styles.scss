@use '@awsui/design-tokens/index.scss' as awsui;

.tm-html-wrapper {
  // DataRow
  .annotation-row {
    white-space: nowrap
  }

  // Rows
  .panel-rows {
    margin-left: 1em;
    margin-right: 1em
  }

  // Container
  .container {
    overflow: auto;
    background-color: awsui.$color-background-container-header;
    border: 1px solid awsui.$color-border-button-normal-disabled;
    border-radius: 2px;
    box-shadow: 0px 1px 4px -2px rgba(0, 28, 36, 0.5);
  }
  .panel-container {
    width: 210px;
    min-height: 2em;
  }
  .annotation-container {
    padding-left: 12px;
    padding-right: 12px;
  }
  .close-button {
    float: right;
    margin-top: 8px;
    margin-right: 8px
  }
  .arrow {
    height: 14px;
    display: flex;
    flex-direction: column;
    align-items: center;

    .arrow-outer {
      width: 18px; 
      height: 18px; 
      position: absolute;
      z-index: -1;
      bottom: 5px;
      transform: rotate(45deg);
    }
    .arrow-inner {
      border: 0px;
      bottom: 6.5px;
      width: 20px; 
      height: 20px; 
      z-index: 1;
    }
  }
}
