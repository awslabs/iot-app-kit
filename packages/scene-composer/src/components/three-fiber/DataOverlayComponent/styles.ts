import { colorBackgroundContainerHeader, colorBorderButtonNormalDisabled } from '@cloudscape-design/design-tokens';
import { CSSProperties } from 'react';

// DataRow
export const tmAnnotationRow: CSSProperties = {
  whiteSpace: 'nowrap',
};

// Rows
export const tmPanelRows: CSSProperties = {
  marginLeft: '1em',
  marginRight: '1em',
};

// Container
export const tmContainer: CSSProperties = {
  overflow: 'auto',
  backgroundColor: colorBackgroundContainerHeader,
  border: `1px solid ${colorBorderButtonNormalDisabled}`,
  borderRadius: '2px',
  boxShadow: '0px 1px 4px -2px rgba(0, 28, 36, 0.5)',
};
export const tmPanelContainer: CSSProperties = {
  paddingTop: '12px',
  paddingBottom: '12px',
  width: '210px',
  minHeight: '2em',
};
export const tmAnnotationContainer: CSSProperties = {
  paddingLeft: '12px',
  paddingRight: '12px',
};
export const tmCloseButtonDiv: CSSProperties = {
  float: 'right',
  marginTop: '8px',
  marginRight: '8px',
};
export const tmCloseButton: CSSProperties = {
  backgroundColor: 'transparent',
  borderColor: 'transparent',
  color: 'white',
};

// Overlay panel arrow
export const tmArrow: CSSProperties = {
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
};
export const tmArrowOuter: CSSProperties = {
  width: '18px',
  height: '18px',
  position: 'absolute',
  zIndex: -1,
  bottom: '-9px',
  transform: 'rotate(45deg)',
};
export const tmArrowInner: CSSProperties = {
  border: '0px',
  bottom: '-7.5px',
  width: '20px',
  height: '20px',
  zIndex: '1',
  boxShadow: 'none',
};
