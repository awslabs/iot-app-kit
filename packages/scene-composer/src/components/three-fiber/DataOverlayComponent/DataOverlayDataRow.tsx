import React, { ReactElement, useCallback, useEffect, useMemo, useState } from 'react';

import { useSceneComposerId } from '../../../common/sceneComposerIdContext';
import { ITwinMakerEntityDataBindingContext } from '../../../interfaces';
import { Component } from '../../../models/SceneModels';
import { useStore } from '../../../store';
import { dataBindingValuesProvider } from '../../../utils/dataBindingUtils';
import { replaceBindingVariables } from '../../../utils/dataBindingVariableUtils';
import { ReactMarkdownWrapper } from '../../wrappers/ReactMarkdownWrapper';
import './styles.scss';

export interface DataOverlayDataRowProps {
  rowData: Component.DataOverlayRow;
  overlayType: Component.DataOverlaySubType;
  valueDataBindings: Component.ValueDataBindingNamedMap[];
}

export const DataOverlayDataRow = ({
  rowData,
  overlayType,
  valueDataBindings,
}: DataOverlayDataRowProps): ReactElement => {
  const sceneComposerId = useSceneComposerId();
  const dataInput = useStore(sceneComposerId)((state) => state.dataInput);
  const dataBindingTemplate = useStore(sceneComposerId)((state) => state.dataBindingTemplate);
  const isEditing = useStore(sceneComposerId)((state) => state.isEditing());

  const [stringContent, setStringContent] = useState<string>('');

  const bindingValuesMap: Record<string, string> = useMemo(() => {
    const result = {};
    valueDataBindings.forEach((binding) => {
      if (!binding.valueDataBinding) {
        return;
      }
      const values: Record<string, unknown> = dataBindingValuesProvider(
        dataInput,
        binding.valueDataBinding,
        dataBindingTemplate,
      );
      result[binding.bindingName] =
        values[(binding.valueDataBinding.dataBindingContext as ITwinMakerEntityDataBindingContext).propertyName];
    });

    return result;
  }, [valueDataBindings, dataInput, dataBindingTemplate]);

  const updateContentWithBindingVariables = useCallback(
    (content: string) => {
      // Show original content in editing mode
      if (isEditing) {
        setStringContent(content);
        return;
      }

      setStringContent(replaceBindingVariables(content, bindingValuesMap));
    },
    [bindingValuesMap],
  );

  useEffect(() => {
    switch (rowData.rowType) {
      case Component.DataOverlayRowType.Markdown: {
        updateContentWithBindingVariables((rowData as Component.DataOverlayMarkdownRow).content);
        break;
      }
      default:
        break;
    }
  }, [bindingValuesMap, rowData]);

  switch (rowData.rowType) {
    case Component.DataOverlayRowType.Markdown: {
      return (
        <ReactMarkdownWrapper
          className={overlayType === Component.DataOverlaySubType.TextAnnotation ? 'annotation-row' : 'panel-row'}
          content={stringContent}
        />
      );
    }
    default:
      return <></>;
  }
};
