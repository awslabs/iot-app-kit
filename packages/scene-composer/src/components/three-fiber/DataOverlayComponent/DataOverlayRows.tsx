import React, { ReactElement } from 'react';

import { Component } from '../../../models/SceneModels';
import { IDataOverlayComponentInternal } from '../../../store/internalInterfaces';

import { DataOverlayDataRow } from './DataOverlayDataRow';
import './styles.scss';

export interface DataOverlayRowsProps {
  component: IDataOverlayComponentInternal;
}

export const DataOverlayRows = ({ component }: DataOverlayRowsProps): ReactElement => {
  return (
    <div
      className={`${
        component.subType === Component.DataOverlaySubType.TextAnnotation ? 'annotation-rows' : 'panel-rows'
      }`}
    >
      {component.dataRows.map((row, index) => {
        return (
          <DataOverlayDataRow
            rowData={row}
            overlayType={component.subType}
            key={index}
            valueDataBindings={component.valueDataBindings}
          />
        );
      })}
    </div>
  );
};
