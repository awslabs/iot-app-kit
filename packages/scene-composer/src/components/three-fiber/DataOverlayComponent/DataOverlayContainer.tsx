import React, { ReactElement, useCallback, useContext, useEffect, useRef, useState } from 'react';
import { Button } from '@awsui/components-react';

import { IDataOverlayComponentInternal, ISceneNodeInternal } from '../../../store/internalInterfaces';
import { Component } from '../../../models/SceneModels';
import { useStore, useViewOptionState } from '../../../store';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';
import useCallbackWhenNotPanning from '../../../hooks/useCallbackWhenNotPanning';

import { DataOverlayRows } from './DataOverlayRows';
import './styles.scss';

export interface DataOverlayContainerProps {
  node: ISceneNodeInternal;
  component: IDataOverlayComponentInternal;
}

export const DataOverlayContainer = ({ component, node }: DataOverlayContainerProps): ReactElement | null => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const containerRef = useRef<HTMLDivElement>(null);

  const selectedSceneNodeRef = useStore(sceneComposerId)((state) => state.selectedSceneNodeRef);
  const setSelectedSceneNodeRef = useStore(sceneComposerId)((state) => state.setSelectedSceneNodeRef);
  const subType = component.subType;

  const componentVisible = useViewOptionState(sceneComposerId).componentVisibilities[subType];
  const initialVisibilitySkipped = useRef(false);

  // TODO: config.isPinned is not supported in milestone 1
  // const [visible, setVisible] = useState(component.config?.isPinned || componentVisible);
  const [visible, setVisible] = useState(componentVisible);

  // Toggle panel visibility on selection change
  useEffect(() => {
    if (selectedSceneNodeRef === node.ref) {
      setVisible(true);
    }
  }, [selectedSceneNodeRef, node.ref]);

  // Toggle visibility on view option change. Skip the first call to make sure the
  // isPinned config can keep panel open initially.
  useEffect(() => {
    if (initialVisibilitySkipped.current) {
      setVisible(componentVisible);
    }
    initialVisibilitySkipped.current = true;
  }, [componentVisible]);

  // Same behavior as other components to select node when clicked on the panel
  const [onPointerDown, onPointerUp] = useCallbackWhenNotPanning(
    (e) => {
      e.stopPropagation();
      if (selectedSceneNodeRef !== node.ref) {
        setSelectedSceneNodeRef(node.ref);
      }
    },
    [selectedSceneNodeRef, node.ref],
  );

  const onClickCloseButton = useCallback(() => {
    setVisible(false);
  }, [setVisible]);

  return visible ? (
    <>
      <div
        ref={containerRef}
        onPointerUp={onPointerUp}
        onPointerDown={onPointerDown}
        className={`container ${
          component.subType === Component.DataOverlaySubType.TextAnnotation ? 'annotation-container' : 'panel-container'
        }`}
      >
        {component.subType === Component.DataOverlaySubType.OverlayPanel && !componentVisible && (
          <div className='close-button'>
            <Button iconName='close' variant='icon' iconAlign='right' onClick={onClickCloseButton} />
          </div>
        )}
        <DataOverlayRows component={component} />
      </div>

      {subType == Component.DataOverlaySubType.OverlayPanel && (
        <div className='arrow'>
          <div className='container arrow-outer' />
          <div className='container arrow-outer arrow-inner' />
        </div>
      )}
    </>
  ) : null;
};
