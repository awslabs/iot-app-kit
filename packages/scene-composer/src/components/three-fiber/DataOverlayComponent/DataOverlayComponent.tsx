import React, { ReactElement, useContext } from 'react';
import { Html } from '@react-three/drei';

import { ISceneNodeInternal } from '../../../store';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';
import { IDataOverlayComponentInternal } from '../../../store/internalInterfaces';

import { DataOverlayContainer } from './DataOverlayContainer';

export interface DataOverlayComponentProps {
  node: ISceneNodeInternal;
  component: IDataOverlayComponentInternal;
}

export const DataOverlayComponent = ({ node, component }: DataOverlayComponentProps): ReactElement => {
  const sceneComposerId = useContext(sceneComposerIdContext);

  return (
    <group>
      <Html
        className='tm-html-wrapper'
        // TODO: add position after finding proper way to always display overlay right above tag
        // position={position}
        style={{
          transform: 'translate3d(-50%,-100%,0)', // make the center of 3D transform the middle of bottom edge
        }}
      >
        <sceneComposerIdContext.Provider value={sceneComposerId}>
          <DataOverlayContainer node={node} component={component} />
        </sceneComposerIdContext.Provider>
      </Html>
    </group>
  );
};
