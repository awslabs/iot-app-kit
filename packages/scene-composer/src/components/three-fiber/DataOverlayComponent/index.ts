import { DataOverlayComponent } from './DataOverlayComponent';

export default DataOverlayComponent;
