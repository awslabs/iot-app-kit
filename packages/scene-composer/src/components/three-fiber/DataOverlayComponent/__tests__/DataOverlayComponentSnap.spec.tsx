import React from 'react';
import { render } from '@testing-library/react';
import { BoxGeometry, Group, Mesh } from 'three';

import { DataOverlayComponent } from '../DataOverlayComponent';
import { Component } from '../../../../models/SceneModels';
import { KnownComponentType } from '../../../../interfaces';
import { IDataOverlayComponentInternal, ISceneNodeInternal, useStore } from '../../../../store';

jest.mock('../DataOverlayContainer', () => ({
  DataOverlayContainer: (...props: unknown[]) => <div data-testid='container'>{JSON.stringify(props)}</div>,
}));

describe('DataOverlayComponent', () => {
  const mockComponent: IDataOverlayComponentInternal = {
    ref: 'comp-ref',
    type: KnownComponentType.DataOverlay,
    subType: Component.DataOverlaySubType.OverlayPanel,
    dataRows: [
      {
        rowType: Component.DataOverlayRowType.Markdown,
        content: 'content',
      },
    ],
    valueDataBindings: [
      {
        bindingName: 'bindingA',
        valueDataBinding: { dataBindingContext: 'dataBindingContext' },
      },
    ],
  };
  const mockNode: Partial<ISceneNodeInternal> = {
    ref: 'node-ref',
    transform: { position: [1, 2, 3], rotation: [0, 0, 0], scale: [0, 0, 0] },
    components: [mockComponent],
  };

  const group = new Group();
  const geometry = new BoxGeometry(2, 4, 6);
  const mesh = new Mesh(geometry);
  group.add(mesh);

  const getObject3DBySceneNodeRef = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();
    useStore('default').setState({ getObject3DBySceneNodeRef });
  });

  it('should render the component correctly', () => {
    const { container } = render(
      <DataOverlayComponent node={mockNode as ISceneNodeInternal} component={mockComponent} />,
    );
    expect(container.getElementsByClassName('tm-html-wrapper').length).toBe(1);
    expect(container).toMatchSnapshot();
  });

  it('should render the component correctly with object 3D size', () => {
    getObject3DBySceneNodeRef.mockReturnValue(group);
    const { container } = render(
      <DataOverlayComponent node={mockNode as ISceneNodeInternal} component={mockComponent} />,
    );
    // expect(container.querySelector('[position="{\\"x\\":0,\\"y\\":1.7,\\"z\\":0}"]')).not.toBeNull();
    expect(container).toMatchSnapshot();
  });
});
