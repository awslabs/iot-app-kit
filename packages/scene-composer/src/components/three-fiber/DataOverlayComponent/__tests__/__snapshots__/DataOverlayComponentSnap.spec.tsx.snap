// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataOverlayComponent should render the component correctly 1`] = `
<div>
  <group>
    <div
      class="tm-html-wrapper"
      data-mocked="Html"
      style="transform: translate3d(-50%,-100%,0);"
    >
      <div
        data-testid="container"
      >
        [{"node":{"ref":"node-ref","transform":{"position":[1,2,3],"rotation":[0,0,0],"scale":[0,0,0]},"components":[{"ref":"comp-ref","type":"DataOverlay","subType":"OverlayPanel","dataRows":[{"rowType":"Markdown","content":"content"}],"valueDataBindings":[{"bindingName":"bindingA","valueDataBinding":{"dataBindingContext":"dataBindingContext"}}]}]},"component":{"ref":"comp-ref","type":"DataOverlay","subType":"OverlayPanel","dataRows":[{"rowType":"Markdown","content":"content"}],"valueDataBindings":[{"bindingName":"bindingA","valueDataBinding":{"dataBindingContext":"dataBindingContext"}}]}},{}]
      </div>
    </div>
  </group>
</div>
`;

exports[`DataOverlayComponent should render the component correctly with object 3D size 1`] = `
<div>
  <group>
    <div
      class="tm-html-wrapper"
      data-mocked="Html"
      style="transform: translate3d(-50%,-100%,0);"
    >
      <div
        data-testid="container"
      >
        [{"node":{"ref":"node-ref","transform":{"position":[1,2,3],"rotation":[0,0,0],"scale":[0,0,0]},"components":[{"ref":"comp-ref","type":"DataOverlay","subType":"OverlayPanel","dataRows":[{"rowType":"Markdown","content":"content"}],"valueDataBindings":[{"bindingName":"bindingA","valueDataBinding":{"dataBindingContext":"dataBindingContext"}}]}]},"component":{"ref":"comp-ref","type":"DataOverlay","subType":"OverlayPanel","dataRows":[{"rowType":"Markdown","content":"content"}],"valueDataBindings":[{"bindingName":"bindingA","valueDataBinding":{"dataBindingContext":"dataBindingContext"}}]}},{}]
      </div>
    </div>
  </group>
</div>
`;
